File: 79. Word Search/solution.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[i].size(); j++) {
                // vector<vector<bool>> visited(board.size(), vector(board[0].size(), false));
                if (dfs(board, word, i, j, 0)) {
                       return true;
                }
            }
        }
        return false;
    }
    
    bool dfs(vector<vector<char>>& board, string& word, int i, int j, int pos) {
        // cout << "word: " << word[pos] << " i: " << i << " j: " << j << " pos: " << pos << endl;
        char c = board[i][j];
        board[i][j] = '#';
        if (word[pos] != c) { board[i][j] = c; return false; }
        if (pos == word.size() - 1) { return true; }
        // left
        if (j - 1 >= 0 && board[i][j -1] != '#' && dfs(board, word, i, j - 1, pos + 1)) {
            return true;
        }
        // top
        if (i - 1 >=0 && board[i - 1][j] != '#' && dfs(board, word, i - 1, j, pos + 1)) {
            return true;
        }
        // right
        if (j + 1 < board[0].size() && board[i][j + 1] != '#' && dfs(board, word, i, j + 1, pos + 1)) {
            return true;
        }
        // bottom
        if (i + 1 < board.size() && board[i + 1][j] != '#' && dfs(board, word, i + 1, j, pos + 1)) {
            return true;
        }
        board[i][j] = c; // For a char that is not in stack, this char can be used later. But the same char cannot be reused.
        return false;
    }
    
    // bool dfs2(vector<vector<char>>& board, string& word, vector<vector<bool>>& visited, int i, int j, int pos) {
    //     // cout << "word: " << word[pos] << " i: " << i << " j: " << j << " pos: " << pos << endl;
    //     visited[i][j] = true;
    //     if (word[pos] != board[i][j]) { visited[i][j] = false; return false; }
    //     if (pos == word.size() - 1) { return true; }
    //     // left
    //     if (j - 1 >= 0 && !visited[i][j -1] && dfs2(board, word, visited, i, j - 1, pos + 1)) {
    //         return true;
    //     }
    //     // top
    //     if (i - 1 >=0 && !visited[i - 1][j] && dfs2(board, word, visited, i - 1, j, pos + 1)) {
    //         return true;
    //     }
    //     // right
    //     if (j + 1 < board[0].size() && !visited[i][j + 1] && dfs2(board, word, visited, i, j + 1, pos + 1)) {
    //         return true;
    //     }
    //     // bottom
    //     if (i + 1 < board.size() && !visited[i + 1][j] && dfs2(board, word, visited, i + 1, j, pos + 1)) {
    //         return true;
    //     }
    //     visited[i][j] = false; // For a char that is not in stack, this char can be used later. But the same char cannot be reused.
    //     return false;
    // }
};

File: 79. Word Search/solution.py

class Solution:
    def dfs(self, board, s, word, x, y, visited):
        visited.add((x, y))
        # print('x: {}, y: {}, s: {}'.format(x, y, s))
        # print(visited)
        if s == word:
            return True
        # left
        if y - 1 >= 0 and (x, y - 1) not in visited and board[x][y - 1] == word[len(s)]:
            if self.dfs(board, s + word[len(s)], word, x, y - 1, visited):
                return True
            else:
                visited.remove((x, y - 1))
        # top
        if x - 1 >= 0 and (x - 1, y) not in visited and board[x - 1][y] == word[len(s)]:
            if self.dfs(board, s + word[len(s)], word, x - 1, y, visited):
                return True
            else:
                visited.remove((x - 1, y))
        # right
        if y + 1 < len(board[x]) and (x, y + 1) not in visited and board[x][y + 1] == word[len(s)]:
            if self.dfs(board, s + word[len(s)], word, x, y + 1, visited):
                return True
            else:
                visited.remove((x, y + 1))
        # bottom
        if x + 1 < len(board) and (x + 1, y) not in visited and board[x + 1][y] == word[len(s)]:
            if self.dfs(board, s + word[len(s)], word, x + 1, y, visited):
                return True
            else:
                visited.remove((x + 1, y))
    """
    @param board: A list of lists of character
    @param word: A string
    @return: A boolean
    """
    def exist(self, board: List[List[str]], word: str) -> bool:
        candidates = []
        for x in range(len(board)):
            for y in range(len(board[x])):
                if board[x][y] == word[0]:
                    candidates.append((word[0], x, y))
        # print(candidates)
        # write your code here
        for candidate in candidates:
            visited = set()
            if self.dfs(board, candidate[0], word, candidate[1], candidate[2], visited):
                return True
        return False
        '''
        ABCE
        SFES
        ADEE
        '''
        